Adding CellsDemo Test Framework and Initial Tests

darth-cheney opened the pull request:
This PR implements automated testing of the interactive CellsDemoPage tests during unit testing.

<!-- Provide a general summary of your changes in the Title above -->

## Motivation and Context
<!-- Why is this change required? -->
When making changes and/or additions to the UI, manually examining all of the CellsDemoPage demos can be a tedious chore. Aside from questions of styling and layout, the basic interactivity of these tests should be amenable to automated testing (once created).
  
We now have an implementation of that ability.
<!-- What problem does it solve? -->
<!--  If it fixes an open issue, please link to the issue here.-->

## Approach
<!-- Describe your changes in reasonable detail -->
The general idea was to use Selenium along with a spun-up instance of the demo webserver to hit the CellsDemoPage pages directly and interact with the elements on the page. As a result, we have a kind of end to end test of the UI and Cells structures.
  
### Selenium and Webserver ###
We are using the python module for [Selenium](https://selenium.dev/), along with the [Chrome WebDriver](https://chromedriver.chromium.org/getting-started). The latter spins up a headless version of Google Chrome that can then be used to query a resulting loaded page through Selenium (and run tests on elements, interaction, data, and results).
  
In order to make this work in our testing framework, we needed to spin up a version of the webtest webserver (and by extension object_db thread) in-code. You can see the implementation [here](https://github.com/APrioriInvestments/object_database/blob/eric-cells_test-rebase/object_database/web/cells_demo/conftest.py).
  
### Fixtures and Test Methods ###
In order to allow test methods to be written across the different CellsDemo files, but to all use the same instance of a running demo server and headless browser, we make use of [Pytest's "fixtures" feature](https://docs.pytest.org/en/latest/fixture.html).
  
With fixtures, we are able to place test methods that use the selenium/server setup throughout the individual cells_demo test files. The configuration for the (single) fixture that we make use of is defined in the `conftest.py` file under `cells_demo`.
  
In addition to coordinating tests across the files, the fixture we use -- called `headless_browser` -- returns an instance of a new object type we've created to ease certain testing methods, called `HeadlessTester`. A single shared instance of this class is sent to all tests using the fixture as a function argument also called `headless_browser` (indeed, this is how pytest "knows" to use the fixture of the same name). As you can see in examples like `buttons.py`, `code_editor.py`, and others, this object is the primary way to interact with the working page.
  
### `tox.ini` and travis config updates ###
We have updated the pytest configuration to collect _all_ files in the `cells_demo/` directory regardless of their suffix. In other words, any file in there will be searched for test methods even if the filename doens't end in `_test.py`.
  
We have also made significant additions to the travis config file. Here we needed to do several things: 1. Download and install a stable version of chrome;
2. Download and install a version of ChromeDriver that is the same as the acquired stable chrome version;
3. Ensure that the two are linked before running the unit tests
4. Install node dependencies and build the JS bundle
  
### Note on JS Tests ###
In the course of updating the Travis config, we noticed that JS tests were not running at all as part of the unit test action. We have added this ability, updated, and all tests are now running and passing.
  
## How Has This Been Tested?
<!-- Describe in reasonable detail how you tested your changes. -->
'This PR is concerned with tests, and therefore everything is test related already.
<!-- If appropriate, include details of your testing environment, -->
<!-- tests ran to see how your change affects other areas of the code, etc. -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

dkrasner committed:
Adding functional CellsDemo selenium tests and travis configuration.

Note that we have also added the previously missing running of
javascript (node) tests during the Travis Unit Test phase.
These are now incorporated and working.

darth-cheney committed:
Fixing Panel bug that was breaking tests

braxtonmckee commented on .travis.yml:
```diff
@@ -48,6 +52,12 @@ jobs:
     - name: "Unit Tests"
       before_script:
+        - wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
+        - sudo apt install ./google-chrome-stable_current_amd64.deb
```
i wonder if we should try pinning these versions. in general, I don't like just using 'latest', because then they can change the version and your tests break and you can't tell whether it was the version change or the commit.

braxtonmckee commented on object_database/web/cells_demo/conftest.py:
```diff
@@ -0,0 +1,270 @@
+"""
```
what's the difference between this and the stuff in CellsDemi_test? looks like duplicate code.

braxtonmckee reviewed (commented):
This looks great. Two points:

1. we need tests to pass before we can merge. looks like a panel test bugging out?
2. I think we should try as much as possible to pin version numbers in the travis yaml file. As long as we document the process for finding new urls with the right versions using comments right next to it, this will prevent us from being broken by an upgrade in the background. of course, if they deprecate a url that would also be bad, so we may have to experiment with this. What do you think? 
3. I saw some duplicated code - we really ought to have only one python file describing booting up all the test infrastructure. I think the conftest one is probably enough...

Otherwise, looks great.

darth-cheney committed:
Removing duplicate file from git

darth-cheney commented:
@braxtonmckee Yo I'll take each point inline:
  
> looks like a panel test bugging out?
  
Yeah, I mentioned this in the email. It looks like JS tests were *not running at all* during travis builds, which I fixed here also. This is why we didn't notice the bug -- introduced with recent panel changes -- until this PR. I have now merged the branch where I took care of the slight panel adjustment into this branch so it should be good to go.
  
> I think we should try as much as possible to pin version numbers in the travis yaml file. 
  
99% of the time I agree: this is the right thing to do. The issue here is that Google does not provide consistent download links for older versions of Chrome -- you can have the beta or the stable. Complicating this matter is that you have to have a Selenium ChromeDriver that matches whatever the Chrome version is. The commands in the travis file take care of that part, but yeah, we can't really specify a version (as far as I've been able to tell) if we want to use "official" Google builds. We can either link from some third party site [like this one](https://www.slimjet.com/chrome/google-chrome-old-version.php) or actually compile older versions, then nab the matching ChromeDriver version. I don't think the juice is worth the squeeze there. But if you really want to we can find some way of making it happen.
  
> I saw some duplicated code
  
Classic case of "removed locally but not in git." Fixed and updated.

braxtonmckee commented:
I wonder if we could use github itself to host the binaries we want to test against? That way we can pick specific versions that we know work, pin them, and rely on github itself to do the serving.

I think for the moment we shouldn't worry about it. But if we start seeing breakages due to version migration under us, I think that picking a specific version and finding our own place to host it should be the strategy...
